src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -122,7 +146,35 @@ public bool TryAcquire(int millisecondsTimeout, bool trackContentions = false)
             return TryAcquireContended(currentThreadId, millisecondsTimeout, trackContentions);
         }
 
-        private bool TryAcquireContended(IntPtr currentThreadId, int millisecondsTimeout, bool trackContentions = false)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal bool TryAcquireOneShot(int currentThreadId)
+        {
+            //
+            // Make one quick attempt to acquire an uncontended lock
+            //
+            if (Interlocked.CompareExchange(ref _state, Locked, Uncontended) == Uncontended)
+            {
+                Debug.Assert(_owningThreadId == 0);
+                Debug.Assert(_recursionCount == 0);
+                _owningThreadId = currentThreadId;
+                return true;
+            }
+
+            return false;
+        }
+
+        private static unsafe void ExponentialBackoff(uint iteration)
+        {
+            if (iteration > 0)
+            {
+                // no need for much randomness here, we will just hash the frame address + iteration.
```

Comment:
This random sequence will often be the same for given thread (the address of the `iteration` will often be same across calls). Can we end up in situation where given thread constantly spins a lot less?

Reply:
Fibonacci hash is very good for scattering periodic sequences (like 1N, 2N, 3N...). The weak spot is if the sequence is already a Fibonacci sequence, but that is not the case for iteration numbers or stack addresses.

Also these backoffs are very rare, so it is good to randomize things, but not a huge deal. 

There are two kinds of contentions that every modern lock needs to deal with:
- one is when we come to the room and pull on the handle and the room is occupied, so we pull on the handle a few more times, but then we leave the phone in the wait list and go to sleep.
This kind of contention will be common for some locks so we detect it by lowering spin count every time the spin is unsuccessful so we do not waste too much time trying to barge into the room.

- the other kind of contention is when multiple entrants try to grab the door handle or put the phone number in the wait list at the same time. We just say "sorry, excuse me" and try again after a delay. We must retry in these cases, but we learn from every collision that contention is worse than we estimated, so we adjust our wait. 
State transitions are fast atomic operations though, and access by multiple threads will be already fairly random. Even intentionally arranging for N threads to come at exactly the same time will be very hard as N grows.

Randomness will help to desynchronize threads a bit to ensure the threads will not behave exactly the same in freaky cases when 100 threads somehow managed to come at the same time, but is not hugely important, and we look primarily at mutual randomness between thread's.

Reply:
I've checked what source of randomness CRITICAL_SECTION uses in this scenario and they use `__rdtsc` timestamp and do not even bother to hash.

Reply:
I am worried about one thread being unlucky and consistency getting unfavorable sequence of spin counts, and other thread being lucky and consistency getting favorable sequence of spin counts. `__rdtsc` does not have that problem.

Reply:
It would be more important for threads to not have the same sequence than ensuring that sequence does not repeat, but overall I think it does not matter a lot as long and synchronized backoffs should be very rare, thus randomness is a "good to have" thing.  
I think the alternative is to not do anything thread-specific. It'll likely be ok too.

Tick count at CPU frequency is a nice way to add some time factor, but we do not have that.  
We have Stopwatch/QPC, but that is typically 10Mhz and at this scale would just produce the same value for all threads involved in the collision, and probably too expensive next to a typical 0-1-2 spins backoff.

Maybe worth thinking about tick counter, if that is portable enough. I think both x64 and arm64 have it. It could be useful for these kind of things and perhaps for benchmarking.

Reply:
Should we set the highest bit to make sure that the backoff is roughly exponential, but still allow the rest of the bits to be random? I think it would address my concern enough.

Reply:
> Should we set the highest bit to make sure that the backoff is roughly exponential, but still allow the rest of the bits to be random?

Yes, I think that will help!